Modularize project

segunfamisa opened the pull request:
This commit extracts existing code into two modules:
1. `floorplan-cli`
2. `core`

This will be useful for reusability and make implementing the Gradle plugin easier to implement - #15

Commit:
Modularize project

This commit extracts existing code into two modules:
1. `floorplan-cli`
2. `core`

This will be useful for reusability.

julioz commented:
While running the tests, I got
```
The Kotlin Gradle plugin was loaded multiple times in different subprojects, which is not supported and may break the build.
This might happen in subprojects that apply the Kotlin plugins with the Gradle 'plugins { ... }' DSL if they specify explicit versions, even if the versions are equal.
Please add the Kotlin plugin to the common parent project or the root project, then remove the versions in the subprojects.
If the parent project does not need the plugin, add 'apply false' to the plugin line.
See: https://docs.gradle.org/current/userguide/plugins.html#sec:subprojects_plugins_dsl
The Kotlin plugin was loaded in the following projects: ':core', ':floorplan-cli'
```

Could you update that bit before we merge?

julioz commented:
Oh, also when trying to run the production code with the following command:
```
./gradlew run --args="samples/tivi-26.json"
```

I got an error:
```
> Task :floorplan-cli:run FAILED
Exception in thread "main" java.io.FileNotFoundException: samples/tivi-26.json (No such file or directory)
        at java.io.FileInputStream.open0(Native Method)
        at java.io.FileInputStream.open(FileInputStream.java:195)
        at java.io.FileInputStream.<init>(FileInputStream.java:138)
        at kotlin.io.FilesKt__FileReadWriteKt.readText(FileReadWrite.kt:125)
        at kotlin.io.FilesKt__FileReadWriteKt.readText$default(FileReadWrite.kt:125)
        at com.zynger.floorplan.AppKt.main(App.kt:16)
```

Maybe we have `path` issues now that the project structure changed. It seems to be something with relative paths, since absolute paths work fine:
```
./gradlew run --args="/Users/juliozynger/code/FloorPlan/samples/tivi-26.json"
```

I gotta jump to another thing right now, but could check the reasons later today if you prefer me to do it.

julioz commented:
By the way, thank you very much for the PR!

julioz commented on floorplan-cli/build.gradle:
```diff
@@ -0,0 +1,37 @@
+    implementation project(":core")
+    implementation "org.jetbrains.kotlin:kotlin-stdlib-jdk8"
+    implementation "org.jetbrains.kotlinx:kotlinx-serialization-runtime:0.20.0" // JVM dependency
+    testCompile group: 'junit', name: 'junit', version: '4.12'
```
Same as above
```suggestion
    testImplementation group: 'junit', name: 'junit', version: '4.12'
```

julioz commented on core/build.gradle:
```diff
@@ -0,0 +1,25 @@
+dependencies {
+    implementation "org.jetbrains.kotlin:kotlin-stdlib-jdk8"
+    implementation "org.jetbrains.kotlinx:kotlinx-serialization-runtime:0.20.0" // JVM dependency
+    testCompile group: 'junit', name: 'junit', version: '4.12'
```
Unrelated, but low-hanging fruit; we could take the chance to update this deprecated configuration with
```suggestion
    testImplementation group: 'junit', name: 'junit', version: '4.12'
```

segunfamisa commented:
Oh interesting. I'll have a look at these.

thanks for the review

Commit:
Use testImplementation instead of testCompile

Commit:
Move version definition of kotlin jvm plugin to root project

This commit fixes the error message below:

> The Kotlin Gradle plugin was loaded multiple times in different subprojects, which is not supported and may break the build.
This might happen in subprojects that apply the Kotlin plugins with the Gradle 'plugins { ... }' DSL if they specify explicit versions, even if the versions are equal.
Please add the Kotlin plugin to the common parent project or the root project, then remove the versions in the subprojects.
If the parent project does not need the plugin, add 'apply false' to the plugin line.
See: https://docs.gradle.org/current/userguide/plugins.html#sec:subprojects_plugins_dsl
The Kotlin plugin was loaded in the following projects: ':core', ':floorplan-cli'

Commit:
Remove repeated ext block

Commit:
Remove unnecessary dependency

segunfamisa commented:
I made some updates to this PR.

Yes, the relative path `samples/tivi-26.json` no longer works because the execution code is no longer at the root of this project. Using `../samples/tivi-26.json` works instead.

An option could be to move the `samples/tivi-26.json` file into `floorplan-cli/samples/tivi-26.json` module since it's a sample specific to the cli feature.

I don't know if it is possible to retain the main class in the root project of a multi-module project while the class physically lives in a subproject

Commit:
Revert "Remove unnecessary dependency"

This reverts commit 18568aac

julioz commented on floorplan-cli/build.gradle:
```diff
@@ -0,0 +1,33 @@
+}
+
+application {
+    mainClassName = javaMainClass
```
Since this is the only gradle script that receives the `application` plugin, I think it would make more sense to inline the value of `javaMainClass` here and remove it from the root `build.gradle` file
```suggestion
    mainClassName = "com.zynger.floorplan.AppKt"
```

segunfamisa replied:
Yeah. This makes sense! 👍 Done in a35bf9b

julioz reviewed (approved):
> Yes, the relative path samples/tivi-26.json no longer works because the execution code is no longer at the root of this project. Using ../samples/tivi-26.json works instead.

Makes sense! I didn't pay attention to that detail with the change in the file-system organization when I first tested. I think this is ok.

> An option could be to move the samples/tivi-26.json file into floorplan-cli/samples/tivi-26.json module since it's a sample specific to the cli feature.

I'd rather have them in the root directory, since these `json` sample files are not specific to the `cli` UI of FloorPlan, theoretically we can use the same ones for any other UI we build. I am ok living with the `../` prefix for them on manual test runs. I also did a quick search on references to absolute paths to `samples/` in the repo and we don't have any (fortunately we always refer to generic `<path-to-schema-file>`) placeholders, so there's no need to update documentation either.

- - - 
Besides the small nit with the `mainClass` definition location, I think this is good to go! Ping me when you can apply the small change there and I'll merge this PR.

Commit:
Remove unnecessary extra property from root project

julioz commented:
Thanks!
